File: GuardianMonsters/core/src/de/limbusdev/guardianmonsters/fwmengine/battle/ui/widgets/BattleWidget.java
package de.limbusdev.guardianmonsters.fwmengine.battle.ui.widgets;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.Touchable;
import com.badlogic.gdx.scenes.scene2d.actions.Actions;
import com.badlogic.gdx.scenes.scene2d.ui.WidgetGroup;

/**
 * For all children of a BattleWidget, callbacks have to be added. When adding a widget to the
 * @link{AHUD}, the method onButtonClicked() can be used, in that case you can enter the button ID
 * into the Enumeration @link{ButtonIDs}
 * Created by georg on 18.08.16.
 */
public abstract class BattleWidget extends WidgetGroup {

    private Runnable runnableRemove;

    public BattleWidget () {
        super();

        // Controller
        runnableRemove = new Runnable() {
            @Override
            public void run() {
                superRemove();
            }
        };

    }

    public void addFadeOutAction(float duration) {
        addAction(Actions.sequence(Actions.alpha(0, duration), Actions.visible(false)));
    }

    public void addFadeInAction(float duration) {
        addAction(Actions.sequence(Actions.visible(true), Actions.alpha(1, duration)));
    }


    public void addToStageAndFadeIn(Stage newParent) {
        clearActions();
        addToStage(newParent);
        addAction(Actions.sequence(Actions.alpha(0),Actions.fadeIn(.5f)));
    }

    public boolean fadeOutAndRemove () {
        if (getParent() != null) {
            addAction(Actions.sequence(Actions.alpha(1), Actions.alpha(0, .3f), Actions.run(runnableRemove)));
            return true;
        }
        else return false;
    }

    public void addToStage(Stage stage) {
        if(getParent() == null) stage.addActor(this);
    }
    private boolean superRemove() {
        return super.remove();
    }

    public void enable() {
        setColor(Color.WHITE);
        setTouchable(Touchable.enabled);
        for(Actor a : getChildren()) {
            enable(a);
        }
    }

    public void disable() {
        setColor(Color.GRAY);
        setTouchable(Touchable.disabled);
        for(Actor a : getChildren()) {
            disable(a);
        }
    }

    public void enable(Actor child) {
        child.setColor(Color.WHITE);
        child.setTouchable(Touchable.enabled);
    }

    public void disable(Actor child) {
        child.setColor(Color.GRAY);
        child.setTouchable(Touchable.disabled);
    }

}

File: GuardianMonsters/battle/src/de/limbusdev/guardianmonsters/battle/ui/widgets/BattleWidget.kt
package de.limbusdev.guardianmonsters.battle.ui.widgets

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.Touchable
import com.badlogic.gdx.scenes.scene2d.actions.Actions.*
import com.badlogic.gdx.scenes.scene2d.actions.Actions.run as runThis
import com.badlogic.gdx.scenes.scene2d.ui.WidgetGroup

import ktx.actors.then


/**
 * BattleWidget adds some convenience methods to [WidgetGroup], like:
 *
 *  + fading in and out
 *  + automated adding and removing of the Widget from and to parent or stage
 *  + enabling and disabling of the widget and all children
 *  + visualizing enabled and disabled states
 *
 * @author Georg Eckert 2019
 */
abstract class BattleWidget : WidgetGroup()
{
    // .................................................................................. Properties
    private val runnableRemove: () -> Unit


    // ................................................................................ Constructors
    init
    {
        // Controller
        runnableRemove = { superRemove() }
    }


    // ..................................................................................... Methods
    /** Smoothly turns the actors alpha to 0 and then makes it invisible. */
    fun addFadeOutAction(duration: Float)
    {
        addAction(alpha(0f, duration) then visible(false))
    }

    /** Makes the actor visible and smoothly turns it's alpha to 1. */
    fun addFadeInAction(duration: Float)
    {
        addAction(visible(true) then alpha(1f, duration))
    }

    /** Clears all actions, adds this actor to the given stage and fades it in. */
    open fun addToStageAndFadeIn(newParent: Stage)
    {
        clearActions()
        addToStage(newParent)
        addAction(alpha(0f) then fadeIn(.5f))
    }

    /** Fades out actor and removes it from it's parent. Returns false, if actor had no parent. */
    open fun fadeOutAndRemove(): Boolean
    {
        parent?.addAction(alpha(1f) then alpha(0f, .3f) then runThis(runnableRemove))

        return (parent != null)
    }

    /** Adds this actor to the given [Stage]. */
    fun addToStage(stage: Stage)
    {
        stage.addActor(this)
    }

    /** Resets the tint (white), makes the actor touchable and enables all children. */
    fun enable()
    {
        color = Color.WHITE
        touchable = Touchable.enabled

        children.forEach{ child -> enable(child) }
    }

    /** Changes the tint to gray, disables touch and disables all children as well */
    fun disable()
    {
        color = Color.GRAY
        touchable = Touchable.disabled

        children.forEach{ child -> disable(child) }
    }

    /** Resets the tint (white) of a child and makes it touchable. */
    fun enable(child: Actor)
    {
        child.color = Color.WHITE
        child.touchable = Touchable.enabled
    }

    /** Changes the child's tint to gray and disables touch for it. */
    fun disable(child: Actor)
    {
        child.color = Color.GRAY
        child.touchable = Touchable.disabled
    }


    // ............................................................................... Super Methods
    /** Makes super's remove() method visible. */
    private fun superRemove(): Boolean = super.remove()
}
